Group exposed ports in ps output

ColmHally opened the pull request:
Fixes #10302 

Signed-off-by: Colm Hally c.hallywebd@gmail.com

jessfraz committed:
Blacklist zfs with overlay

Docker-DCO-1.1-Signed-off-by: Jessica Frazelle <jess@docker.com> (github: jfrazelle)

thaJeztah commented:
Thanks, @ColmHally

An example output was posted here: https://github.com/docker/docker/issues/10302#issuecomment-71396684

Should the output in https://docs.docker.com/reference/commandline/cli/#ps be updated to include a range (for illustration)? A short note below the example describing that port-ranges are shown "grouped" would be nice there as well.

ColmHally commented:
Thanks, @thaJeztah 
Good catch re: docs - I'll add a short example now to the docs.

thaJeztah commented:
I think there's no real need to add a "separate" output, but it can be put in the output that's already there.

ColmHally commented:
@thaJeztah, I've added a short note to the docs and rebased my commits. Any further thoughts?

Also, I didn't realise that it was also possible to publish a range of ports to the host.. Do you think we need to add support for those ranges in the `ps` output?

thaJeztah commented:
If possible, yes. I think the original issue also was about publishing a large range of ports (e.g 1000-2000)

SvenDowideit commented:
Docs LGTM - @fredlf @jamtur01

jamtur01 commented:
Build failed.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1434,12 +1434,14 @@ Running `docker ps` showing 2 linked containers.
 `docker ps` will show only running containers by default. To see all containers:
 `docker ps -a`
 
+`docker ps` will group exposed ports into a single range if possible, e.g. a container that exposes TCP ports `100, 101, 102` will display `100-102/tcp` in the `PORTS` column.
```
Run-on. Need to break this into two sentences starting at E.g. Need a comma after the "g." as well, please!

fredlf commented:
Docs LGTM after my copy edit. Thanks for the contribution.

ColmHally commented:
Thanks @fredlf - will incorporate your edits and add support for published ports later this evening.

ColmHally committed:
Group published and exposed ports by contiguous ranges

Signed-off-by: Colm Hally <colmhally@gmail.com>

ColmHally commented:
I've added support for grouping published ranges of ports and @fredlf, I've made the changes you pointed out as well.

Grouping will work for ports published like so:
```
$ docker run -p 1000-2000:1000-2000 ubuntu:14.40 bash
eefa3dbcdd1cba22540dcf6787d7d2c56d654dc95b693de7027d4f35ed889cc0
$ docker ps
eefa3dbcdd1c        ubuntu:14.04        "bash"              11 seconds ago      Up 5 seconds        0.0.0.0:1000-2000->1000-2000/tcp   modest_galileo 
```
where there is an exact match between the host and internal container port range. However, ports published without an exact match (such as `-p 1500-2500:1000-2000` or `-p :1500-2500`) are simply printed out individually (as before):

```
$ docker run -p 1500-2500:1000-2000 ubuntu:14.40 bash
5cebbad0bf0cf5a1ab482fd22b889b04f61851045ceabfd9a70d46dec0527365
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS    NAMES
5cebbad0bf0c        ubuntu:14.04        "bash"              2 minutes ago       Up About a minute   0.0.0.0:1500->1000/tcp, 0.0.0.0:1501->1001/tcp, 0.0.0.0:1502->1002/tcp, 0.0.0.0:1503->1003/tcp, 0.0.0.0:1504->1004/tcp, 0.0.0.0:1505->1005/tcp, 0.0.0.0:1506->1006/tcp, 0.0.0.0:1507->1007/tcp, 0.0.0.0:1508->1008/tcp, 0.0.0.0:1509->1009/tcp, 0.0.0.0:1510->1010/tcp, 0.0.0.0:1511->1011/tcp, 0.0.0.0:1512->1012/tcp, 0.0.0.0:1513->1013/tcp, 0.0.0.0:1514->1014/tcp, 0.0.0.0:1515->1015/tcp, 0.0.0.0:1516->1016/tcp, 0.0.0.0:1517->1017/tcp, 0.0.0.0:1518->1018/tcp ...............
```

Another more comprehensive example:
```
$ docker run -dt -p 443:443 --expose=400-405/udp -p 4327-4655:4327-4655 -p 127.0.0.1:6000-6003:6000-6003 -p 127.0.0.1:32000-32005:32000-32005/udp ubuntu:14.04 bash
519be5a7927fb7ac40ca7550216c623f3737d5f216d58e36be7ff61d4718b92e
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS              PORTS                                                                                                                                             NAMES
519be5a7927f        ubuntu:14.04        "bash"              About a minute ago   Up 10 seconds       0.0.0.0:443->443/tcp, 400-405/udp, 0.0.0.0:4327-4655->4327-4655/tcp, 127.0.0.1:6000-6003->6000-6003/tcp, 127.0.0.1:32000-32005->32000-32005/udp   serene_torvalds     
```

LK4D4 commented:
@ColmHally I noticed that you have unrelated commit from @jfrazelle in your PR. Could you remove it pls?

LK4D4 commented:
@ColmHally Also, could you pls write test in `integration-cli`. Should be pretty simple.

LK4D4 commented on api/common.go:
```diff
@@ -29,21 +29,78 @@ func ValidateHost(val string) (string, error) {
+		var (
+			current      = port.GetInt("PrivatePort")
+			portKey      = port.Get("Type")
+			startOfGroup int
```
pretty confusing naming, should `start/end` or `first/last`

jfrazelle commented:
hey @ColmHally looks like this needs a rebase, feel free to ping me right after and I will review, sorry for the delay on this, we really appreciate this patch :)

jfrazelle commented:
hey @ColmHally I ended up rebasing for you and adding a test in #10835 so you don't need to worry about back and forth with fixes now. Thanks so much for the patch.

ColmHally commented:
Thanks @jfrazelle and @LK4D4! Have been meaning to get to this, but just didn't get a chance.
